Disable foreign key checks on purge for all mysql versions

croensch opened the pull request:
For your consideration, see #272. I don't know if the discussion was a yes/no to this feature.

I can push a little refactor of `getTruncateTableSQL()` to make it fit better.

robertfausk committed:
disable foreign key checks on purge for all mysql versions

... when purge mode is truncate

* add test for change of foreign key check when using mysql
* add test for not set foreign key check when using sqlite

greg0ire commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -280,4 +282,22 @@ private function getDeleteFromTableSQL(string $tableName, AbstractPlatform $plat
         return 'DELETE FROM ' . $tableIdentifier->getQuotedName($platform);
     }
+
+    /**
+     * @param AbstractPlatform $platform
+     * @param Connection       $connection
+     * @param string           $tbl
+     *
+     * @return string
+     */
+    private function getTruncateTableSQL(AbstractPlatform $platform, Connection $connection, $tbl)
```
```suggestion
    private function getTruncateTableSQL(AbstractPlatform $platform, Connection $connection, string $tbl): string
```

greg0ire commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -280,4 +282,22 @@ private function getDeleteFromTableSQL(string $tableName, AbstractPlatform $plat
+    {
+        $sql = $platform->getTruncateTableSQL($tbl, true);
+
+        if ($connection->getDriver() instanceof AbstractMySQLDriver) {
```
A comment explaining in detail why this is needed would be nice.

croensch replied:
You mean a comment in the code or just here?

I have a test which relies on the auto-incremented ids of a small number of fixtures. I can't purge with delete because that does not reset the auto-increment to 1. I can't purge with truncate, which would reset the auto-increment, because that does not work on MySQL once i have a single relation mapped. In both cases i can only drop & create the database before loading the fixtures again.

greg0ire replied:
A code comment could be great.

> because that does not work on MySQL once i have a single relation mapped

Why does it not work? That's what the comment should explain, possibly with references to the MySQL docs.

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,81 @@
+/**
+ * @author Robert Freigang <robertfreigang@gmx.de>
+ *
+ * @covers ORMPurger::getTruncateTableSQL
```
From https://phpunit.readthedocs.io/en/9.5/annotations.html?highlight=covers#covers:

> Please note that this annotation requires a fully-qualified class name (FQCN). To make this more obvious to the reader, it is recommended to use a leading backslash (even if this not required for the annotation to work correctly).

Also, the table in that documentation says that it is not recommended to specify which method is meant to be covered.

croensch replied:
Sorry, i forgot running the added test after rebasing. Anyway i fixed it up now.

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    /**
+     * @return array
```
```suggestion
     * @return list<array{Driver, bool}>
```

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,81 @@
+    const FOREIGN_KEY_CHECK_STRING_END = ';SET FOREIGN_KEY_CHECKS = 1;';
+    const TEST_TABLE_NAME = 'test_table_name';
+
+    /**
+     * @param Driver $driver
+     * @param bool   $hasForeignKeyCheckString
+     *
+     * @dataProvider purgeForDifferentDriversProvider
+     */
+    public function testPurgeForDifferentDrivers(Driver $driver, $hasForeignKeyCheckString)
```
```suggestion
    /**
     * @dataProvider purgeForDifferentDriversProvider
     */
    public function testPurgeForDifferentDrivers(Driver $driver, bool $hasForeignKeyCheckString): void
```

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,81 @@
+        ];
+    }
+
+    /**
+     * @param Driver $driver
+     *
+     * @return string
+     */
+    private function getTruncateTableSQLForDriver(Driver $driver)
```
```suggestion
    private function getTruncateTableSQLForDriver(Driver $driver): string
```

croensch committed:
fix types and original test

croensch committed:
fix types

as suggested

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

croensch committed:
fix cs errors

croensch committed:
add comment similar to mysql docs

croensch commented:
I initially picked up and rebased this PR to revive _or_ end the discussion around this problem. Now that i looked into it, i understand a few things better:

- The ORMPurger touches all tables anyway, so for MySQL where the TRUNCATE is a drop/re-create internally, it should theoretically not be much faster than the current workaround of running the drop & create commands. Still, it would be nice if this feature just works on MySQL too.
- In AliceDataFixtures they implemented a fix referencing the original PR. We should not break that, i don't think it does, but it got me thinking... https://github.com/theofidry/AliceDataFixtures/blob/master/src/Bridge/Doctrine/Purger/Purger.php
- So i like how DBUnit restores the variables state, perfectly fine pattern to be least-invasive. https://github.com/sebastianbergmann/dbunit/blob/master/src/Operation/Truncate.php
- In DBUnit but also this PR i don't like how it's done around every TRUNCATE statement as it may spam the logger. I would rewrite it to call a method (like in DBUnit) before and after the tables loop.
- I still think it's a good idea to sniff the connection instead of the platform since only `AbstractMySQLDriver` is the same since 3.0 while the `AbstractMySQLPlatform` is only available since DBAL 3.3, but `3.0` is the minimum dependency.

So would you mind if i rewrite it to address all this? In the same PR?

greg0ire commented:
> DBAL 3.3, but 3.0 is the minimum dependency.

I think it's fine to bump the minimum dependency to 3.3, sniffing the ~connection~ driver (why did you mention the connection?) is not a great idea because of middlewares that can wrap the driver in an arbitrary number of decorators.

> So would you mind if i rewrite it to address all this? In the same PR?

Sure, go ahead :slightly_smiling_face:

croensch committed:
refactor completely

- sniff platform instead of driver
- restore variable in MySQL session
- only enable/disable once per purge

croensch committed:
restore CASCADE

greg0ire commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -280,4 +288,38 @@ private function getDeleteFromTableSQL(string $tableName, AbstractPlatform $plat
         return 'DELETE FROM ' . $tableIdentifier->getQuotedName($platform);
     }
+
+    private function disableForeignKeyChecksForMySQL(Connection $connection)
```
```suggestion
    private function disableForeignKeyChecksForMySQL(Connection $connection): void
```

greg0ire commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -280,4 +288,38 @@ private function getDeleteFromTableSQL(string $tableName, AbstractPlatform $plat
+        $connection->executeStatement('SET FOREIGN_KEY_CHECKS = 0');
+    }
+
+    private function enableForeignKeyChecksForMySQL(Connection $connection)
```
```suggestion
    private function enableForeignKeyChecksForMySQL(Connection $connection): void
```

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,103 @@
+    public const TEST_CLASS_NAME = Link::class;
+    public const TEST_TABLE_NAME = 'link';
+
+    /** @return MappingDriver */
```
Why not use a native return type hint here?

croensch replied:
I was doing it like the BaseTest::getMock*() functions. No return type, but hinting the real class.

I guess like ORMExecutorTest::getMockFixture() is the way to go?

greg0ire replied:
Yes, because `BaseTest::getMock*()` do not return _phpunit_ mocks, unlike what you are doing.

croensch replied:
Indeed, i got confused there. Solved now.

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,103 @@
+        $purger->purge();
+    }
+
+    /** @return list<array{AbstractPlatform, int, bool}> */
```
```suggestion
    /** @return list<array{AbstractPlatform, ORMPurger::PURGE_MODE_*, bool}> */
```

croensch replied:
Cool, did not know.

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,103 @@
+        return $metadataDriver;
+    }
+
+    /** @return Connection */
+    protected function getMockConnectionForPlatform(AbstractPlatform $platform)
```
```suggestion
    /** @return Connection&MockObject */
    protected function getMockConnectionForPlatform(AbstractPlatform $platform): Connection
```

croensch replied:
Yes! Let me figure out how i can add the change to the previous function via this UI.

croensch commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,103 @@
+    public const TEST_CLASS_NAME = Link::class;
+    public const TEST_TABLE_NAME = 'link';
+
+    /** @return MappingDriver */
+    protected function getMockMetadataDriver()
```
```suggestion
    /** @return MappingDriver&MockObject */
    protected function getMockMetadataDriver(): MappingDriver
```

croensch commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -5,14 +5,18 @@
 use Doctrine\DBAL\Schema\Identifier;
 use Doctrine\ORM\EntityManagerInterface;
 use Doctrine\ORM\Mapping\ClassMetadata;
 
```
```suggestion
use PHPUnit\Framework\MockObject\MockObject;

```

croensch committed:
fix types and hints

as suggested

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

croensch committed:
fix mistake

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,104 @@
+        $metadataDriver = $this->createMock(MappingDriver::class);
+        $metadataDriver->method('getAllClassNames')->willReturn([self::TEST_CLASS_NAME]);
+        $metadataDriver->method('loadMetadataForClass')
+            ->willReturnCallback(static function (string $className, ClassMetadata $metadata) {
```
```suggestion
            ->willReturnCallback(static function (string $className, ClassMetadata $metadata): void {
```

greg0ire commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /** @return list<array{AbstractPlatform, ORMPurger::PURGE_MODE_*, bool}> */
+    public function purgeForDifferentPlatformsProvider()
```
```suggestion
    public function purgeForDifferentPlatformsProvider(): array
```

croensch committed:
apply suggestions on test

Co-authored-by: Grégoire Paris <postmaster@greg0ire.fr>

croensch commented:
Sorry for the spam. I got nothing to add or change - it's finished for me.

Regarding the DBAL (Common?) dependency i also check the previous class in a few easy-to-delete lines. Just in case you want to postpone upping the dependency.

greg0ire commented:
Reviewers: to be squash-merged

croensch commented:
@SenseException @derrabus can we get another look at this?

SenseException reviewed (approved):
Having a functional test with the foreign keys would be a nice-to-have, even though I don't expect breaking changes in the methods.

derrabus commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -280,4 +288,38 @@ private function getDeleteFromTableSQL(string $tableName, AbstractPlatform $plat
+        $platform = $connection->getDatabasePlatform();
+
+        if (! class_exists(AbstractMySQLPlatform::class)) {
+            // before DBAL 3.3
```
We don't need to support ancient DBAL versions anymore.

derrabus commented on lib/Doctrine/Common/DataFixtures/Purger/ORMPurger.php:
```diff
@@ -163,6 +169,8 @@ public function purge()
             }
         }
+
+        $this->enableForeignKeyChecksForMySQL($connection);
```
If anything inside the foreach throws, we leave the connection in a state of disabled foreign key checks, don't we? We should do something about that.

derrabus commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,104 @@
+    public const TEST_TABLE_NAME = 'link';
+
+    /** @return MappingDriver&MockObject */
+    protected function getMockMetadataDriver(): MappingDriver
```
```suggestion
    private function getMockMetadataDriver(): MappingDriver
```

derrabus commented on tests/Doctrine/Tests/Common/DataFixtures/Purger/ORMPurgerForeignKeyCheckTest.php:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /** @return Connection&MockObject */
+    protected function getMockConnectionForPlatform(AbstractPlatform $platform): Connection
```
```suggestion
    private function getMockConnectionForPlatform(AbstractPlatform $platform): Connection
```

derrabus reviewed (commented):
The 1.5.x branch is closed for features. We need to target 1.7.x now.
